nearcore/src/state_sync.rs

```diff
@@ -304,6 +328,249 @@ async fn state_sync_dump(
     tracing::debug!(target: "state_sync_dump", shard_id, "Stopped state dump thread");
 }
 
+async fn get_missing_state_parts_for_epoch(
+    shard_id: ShardId,
+    chain_id: &String,
+    epoch_id: &EpochId,
+    epoch_height: u64,
+    total_parts: u64,
+    external: &ExternalConnection,
+) -> Result<Vec<u64>, anyhow::Error> {
+    let s3_directory =
+        external_storage_location_directory_multi_node(chain_id, epoch_id, epoch_height, shard_id);
+    let file_names = external.list_state_parts(shard_id, &s3_directory).await?;
+    let mut res = vec![];
+    if !file_names.is_empty() {
+        let mut total_parts: u64 = 0;
+        let mut existing_nums = HashSet::new();
+        for name in file_names {
+            let splitted: Vec<_> = name.split("_").collect();
+            let part_id = splitted.get(2).unwrap().to_string().parse::<u64>()?;
+            if total_parts == 0 {
+                total_parts = splitted.get(4).unwrap().to_string().parse::<u64>()?;
+            }
+            existing_nums.insert(part_id);
+        }
+        for i in 0..total_parts {
+            if !existing_nums.contains(&i) {
+                res.push(i)
+            }
+        }
+        let num_missing = res.len();
+        tracing::debug!(target: "state_sync_dump", ?num_missing, ?s3_directory, "number of missing parts: ");
+        Ok(res)
+    } else {
+        for i in 0..total_parts {
```

Comment:
This loop is duplicated. Please consider restructuring it in a manner similar to the following:
```
let existing_nums : HashSet<_> = file_names.map(|file_name| smth-smth).collect();
let missing_nums = 0..total_parts.filter(|i| !existing_sums.contains(i)).collect();
```